### ISSUE TITLE
[discussion] In-place gradient (grad_input) computation for better memory utilisation

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#46168]
grad_input can sometimes be computed inplace, reuse grad_output and not allocate another buffer (this is often the case for normalization or elementwise ops). This would lead to better memory utilisation.

Originally discussed in: https://github.com/pytorch/pytorch/issues/26288#issuecomment-663731698

@albanD's argument was that this may break hooks. I propose to allow some ways for user to explicitly guarantee that no backward hook registration is supported for some parts of the model, and to even detect automatically if any hooks are registered. and allow this mode.

**Comment by albanD** [comment]
Note that this comment not only mentions hook but also all the views that can happen between different grads in the network.
And because we don't track all the other Tensors that are views of this one, we cannot know if other Tensors are looking at the same memory as this one. And so I don't think we can expect to know when it is safe to modify such Tensor inplace.

That being said it would be very interesting indeed. So very curious if you have ideas around this?

**Comment by ngimel** [comment]
A common case where the gradients are views is `z=x+y`. In the absence of broadcasting, gradients of x and y would be referencing the same storage, and then would go their own separate ways, so modifying them in place would be unsafe. 
Also note that the memory high water mark is typically achieved at the very beginning of the backward pass, so for most of backward it does not matter much whether gradients are inplace or out-of-place (although it can lead to less fragmentation).

**Comment by vadimkantorov** [comment]
Yep, it's mostly fragmentation effects that I'm talking about coming from absence of fine-grained controls over memory allocator buffer bucketing (and is it also faster if we read/write the same memory from some caches standpoint?)

In BatchNorm + activation "single-stream" architectures with custom backwards that currently just allocate a new grad_input, whereas they could reuse grad_output if it's known to not be used by anything else and is of good size.

**Comment by albanD** [comment]
I think that your comment actually points in the right direction: if you have specific group of ops that you know will behave well and you want to optimize, then you can wrap them in a single `autograd.Function`. That way you might be able to save even more buffers and you can use the intermediary gradients inplace as they are internal to your Function.

**Comment by vadimkantorov** [comment]
Maybe it's possible to have some sort of MemoryEconomicalSequential

**Comment by vadimkantorov** [comment]
And enabling manual shape hints / shape bucketing would also help of course :)

**Comment by albanD** [comment]
> Maybe it's possible to have some sort of MemoryEconomicalSequential (not supporting hooks etc) that could be wrapped into a single autograd function

But then you will still need to implement the backward for each op as an inplace op (which we don't have right now).
Note also that you will need extra logic to only save the minimal set of buffers (which is handled by the autograd but you won't have that for the intermediary things inside your custom Function).
So in spirit you want something like that. But I don't think you can implement a generic one easily. It will have to be specialized for a group of ops.

**Comment by albanD** [comment]
> refcount to detect absence of other views 

We don't have that I'm afraid :/
refcount just know how many reference to this Tensor object they are. Not how many Tensor look at the same memory as this Tensor.

**Comment by ngimel** [comment]
Storages have refcounts too, I think currently this is used to steal produced gradients to move them (instead of copying) to `.grad` attributes.

**Comment by vadimkantorov** [comment]
In many cases allocation of grad_input is already separated from the function that fills in the buffer. In many cases (elementwise functions, normalization) passing in same grad_input pointer as grad_output pointer would just work.

**Comment by albanD** [comment]
> Storages have refcounts too, I think currently this is used to steal produced gradients to move them (instead of copying) to .grad attributes.

I don't think we do no. We just look at the Tensor refcount.
And it is expected that the returned Tensors can (both for .backward and autograd.grad), in some rare cases, be views of each other.
For example, taking your example from above:
```
>>> a, b = torch.rand(10,requires_grad=True), torch.rand(10,    requires_grad=True)
>>> c = (a+b) * 2 # multiplication to make the grad contiguous. Otherwise, it can't be re-used
>>> c.sum().backward()
>>> b.grad.data_ptr()
140321091717824
>>> b.grad.data_ptr()
140321091717824
```

**Comment by vadimkantorov** [comment]
Maybe one way would be to extend `backward()` or `autograd.grad` by enabling them to take an `out` argument that would somehow contain the buffers (mainly for grad_input, non-leaf nodes). There is a question on how to name the buffers. Then if there's not too much overhead, these methods could be used by a implementations of backprop of custom memory-efficient containers

**Comment by albanD** [comment]
Well `backward()` will already re-use the `.grad` field of the leaf Tensors if it exists. So it already behaves as if you had this `out` argument.
Not sure what you mean by other buffers...

**Comment by vadimkantorov** [comment]
But not for gradInput, right? since these are intermediate and not leaf tensors

**Comment by albanD** [comment]
Not for gradInputs no. We don't know how these are going to be used so we kind of have to create new ones indeed.

**Comment by vadimkantorov** [comment]
Yep, so maybe an intermediate step for more control could be accepting some sort of `out=` argument for autograd.grad

**Comment by ezyang** [comment]
cc @bdhirsh maybe functionalization can help here
